proposals/0366-enforce-data-complete-shred-placement.md

```diff
@@ -0,0 +1,237 @@
+---
+simd: '0366'
+title: Enforce DATA_COMPLETE_SHRED Placement
+authors:
+  - ksn6 (Anza)
+  - Ashwin Sekar (Anza)
+category: Standard
+type: Core
+status: Review
+created: 2025-09-21
+feature: https://github.com/anza-xyz/agave/pull/8099
+---
+
+## Summary
+
+This SIMD enforces that the `DATA_COMPLETE_SHRED` flag can only appear on the
+final data shred within an FEC (Forward Error Correction) set. One key
+use-case for this SIMD: this restriction enables efficient detection of
+soon-to-be-introduced `BlockComponent`s at shred ingestion time, providing
+performance improvements for critical consensus operations. The alternative to
+this SIMD would detect `BlockComponent`s during replay, which would run
+substantially slower. One key example involves `UpdateParent` detection in
+Alpenglow's fast leader handover; detection during replay would not work well,
+as malicious leaders could intentionally misplace `DATA_COMPLETE_SHRED` flags
+to force validators into expensive search operations, creating DoS attack
+vectors and delaying block repairs.
+
+## Motivation
+
+Currently, the `DATA_COMPLETE_SHRED` flag can theoretically appear on any data
+shred within an FEC set, though in practice it marks FEC set boundaries. This
+ambiguity creates significant performance challenges:
```

Comment:
Could also note that there is extra code complexity in storing the data complete ranges for deserialization in replay. With this change and SIMD-0317 (fixed 32:32 FEC sets), it becomes really easy to figure out the ranges to deserialize for replay.